src/sentry/static/sentry/app/components/avatarCropper.tsx

```diff
@@ -0,0 +1,464 @@
+      }
+      if (editingLeft) {
+        newDimensions.left = newDimensions.left + newDimensions.size - this.MIN_DIMENSION;
+      }
+      newDimensions.size = this.MIN_DIMENSION;
+    }
+
+    return {...oldDimensions, ...newDimensions};
+  };
+
+  validateImage() {
+    const img = this.image.current;
+
+    if (!img) {
+      return null;
+    }
+
+    if (img.naturalWidth < this.MIN_DIMENSION || img.naturalHeight < this.MIN_DIMENSION) {
+      return tct('Please upload an image larger than [size]px by [size]px.', {
+        size: this.MIN_DIMENSION - 1,
+      });
+    }
+
+    if (img.naturalWidth > this.MAX_DIMENSION || img.naturalHeight > this.MAX_DIMENSION) {
+      return tct('Please upload an image smaller than [size]px by [size]px.', {
+        size: this.MAX_DIMENSION,
+      });
+    }
+
+    return null;
+  }
+
+  drawToCanvas() {
+    const canvas = this.canvas.current;
+    if (!canvas) {
+      return;
+    }
+
+    const image = this.image.current;
+    if (!image) {
+      return;
+    }
+
+    const {left, top, size} = this.state.resizeDimensions;
+    // Calculate difference between natural dimensions and rendered dimensions
+    const ratio =
+      (image.naturalHeight / image.clientHeight +
+        image.naturalWidth / image.clientWidth) /
+      2;
+    canvas.width = size * ratio;
+    canvas.height = size * ratio;
+
+    canvas
+      .getContext('2d')!
+      .drawImage(
+        image,
+        left * ratio,
+        top * ratio,
+        size * ratio,
+        size * ratio,
+        0,
+        0,
+        size * ratio,
+        size * ratio
+      );
+
+    this.props.updateDataUrlState({dataUrl: canvas.toDataURL()});
+  }
+
+  get imageSrc() {
+    const {savedDataUrl, model, type} = this.props;
+    const uuid = model && model.avatar.avatarUuid;
+    const photoUrl = uuid && `/${AVATAR_URL_MAP[type] || 'avatar'}/${uuid}/`;
+
+    return savedDataUrl || this.state.objectURL || photoUrl;
+  }
+
+  uploadClick = (ev: React.MouseEvent<HTMLAnchorElement>) => {
+    ev.preventDefault();
+    this.file.current && this.file.current.click();
+  };
+
+  renderImageCrop() {
+    const src = this.imageSrc;
+    if (!src) {
+      return null;
+    }
+
+    const {resizeDimensions, resizeDirection} = this.state;
+    const style = {
+      top: resizeDimensions.top,
+      left: resizeDimensions.left,
+      width: resizeDimensions.size,
+      height: resizeDimensions.size,
+    };
+
+    return (
+      <ImageCropper resizeDirection={resizeDirection}>
+        <CropContainer innerRef={this.cropContainer}>
+          <img
+            ref={this.image}
+            src={src}
+            onLoad={this.onImageLoad}
+            onDragStart={e => e.preventDefault()}
+          />
+          <Cropper style={style} onMouseDown={this.onMouseDown}>
+            {Object.keys(resizerPositions).map(pos => (
+              <Resizer
+                key={pos}
+                position={pos as Position}
```

Comment:
@EvanPurkhiser I'm confused why this type cast is necessary

Reply:
Yeah I am too

Reply:
@EvanPurkhiser What's the implicit type of `pos` normally? And what's the error you get?

Reply:
string, `position` prop needs Position. `Object.keys` isn't smart.

Reply:
@EvanPurkhiser Yea, I pulled down for myself. I wonder if there is a more TS friendly way of doing `Object.keys`. But the PR looks good.